File: velocity/src/main/java/ru/arc/velocity/listeners/JoinListener.java
package ru.arc.velocity.listeners;

import com.velocitypowered.api.event.Subscribe;
import com.velocitypowered.api.event.connection.DisconnectEvent;
import com.velocitypowered.api.event.connection.PostLoginEvent;
import com.velocitypowered.api.event.player.ServerConnectedEvent;
import com.velocitypowered.api.proxy.Player;
import com.velocitypowered.api.proxy.ProxyServer;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import net.kyori.adventure.text.Component;
import ru.arc.CommonCore;
import ru.arc.config.Config;
import ru.arc.discord.DiscordBot;
import ru.arc.velocity.Velocity;
import ru.arc.xserver.JoinMessages;

import java.util.Optional;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.TimeUnit;

import static ru.arc.Utils.mm;

@Slf4j
public record JoinListener(CommonCore commonCore, ProxyServer proxyServer, Config config) {

    @Subscribe(async = true)
    public void onPlayerJoin(PostLoginEvent event) {
        if (Velocity.isShuttingDown.get()) return;
        proxyServer.getScheduler()
                .buildTask(Velocity.plugin, () -> joinMessage(event.getPlayer()))
                .delay(1, TimeUnit.SECONDS)
                .schedule();
        Optional<String> server = event.getPlayer().getCurrentServer().map(o -> o.getServerInfo().getName());
        commonCore.getPlayerListAnnouncer().addPlayer(event.getPlayer().getUniqueId(), event.getPlayer().getUsername(), server.orElse(null));
    }

    @Subscribe(async = true)
    public void onPlayerLeave(DisconnectEvent event) {
        if (!event.getPlayer().hasPermission("arc.join-message.leave")) return;
        if (Velocity.isShuttingDown.get()) return;
        proxyServer.getScheduler()
                .buildTask(Velocity.plugin, () -> leaveMessage(event.getPlayer()))
                .delay(1, TimeUnit.SECONDS)
                .schedule();
        commonCore.getPlayerListAnnouncer().removePlayer(event.getPlayer().getUniqueId());
    }

    @Subscribe(async = true)
    public void onChangeServer(ServerConnectedEvent event) {
        String server = event.getServer().getServerInfo().getName();
        String username = event.getPlayer().getUsername();
        commonCore.getPlayerListAnnouncer().updatePlayer(event.getPlayer().getUniqueId(), username, server);
    }

    private void sendMessageToAll(Component component) {
        proxyServer.getAllPlayers().forEach(p -> p.sendMessage(component));
    }

    private void joinMessage(Player player) {
        if (Velocity.isShuttingDown.get()) return;
        try {
            if (!player.isActive()) return;
            boolean firstTime = commonCore.getFirstJoinData().firstTimeJoin(player.getUsername());
            if (firstTime) {
                commonCore.getFirstJoinData().markAsJoined(player.getUsername());
                if (!player.hasPermission("arc.join-message.first")) return;
                commonCore.getDiscordBot().sendJoinEmbed(player.getUsername(), DiscordBot.JoinType.FIRST_TIME, null);
                commonCore.getTelegramBot().sendJoinMessage(player.getUsername(), DiscordBot.JoinType.FIRST_TIME, null);
                String message = config.string("messages.first-join", "<gray>Игрок <green>%player_name% <gray>присоединился впервые!");
                message = message.replace("%player_name%", player.getUsername());
                message = config.string("messages.join-prefix", "<dark_green>❖ ") + message;
                sendMessageToAll(mm(message));
            } else {
                if (!player.hasPermission("arc.join-message.join")) return;
                String messageOverride = null;
                if (commonCore.getJoinMessagesRedisRepo() != null) {
                    JoinMessages join = commonCore.getJoinMessagesRedisRepo().getOrNull(player.getUsername()).join();
                    if (join != null && !join.getJoinMessages().isEmpty()) {
                        int idx = ThreadLocalRandom.current().nextInt(join.getJoinMessages().size());
                        int current = 0;
                        for (String joinMessage : join.getJoinMessages()) {
                            if (current == idx) {
                                messageOverride = joinMessage;
                                messageOverride = messageOverride.replace("%player_name%", player.getUsername());
                                break;
                            }
                            current++;
                        }
                    }
                }
                commonCore.getDiscordBot().sendJoinEmbed(player.getUsername(), DiscordBot.JoinType.JOIN, messageOverride);
                commonCore.getTelegramBot().sendJoinMessage(player.getUsername(), DiscordBot.JoinType.JOIN, messageOverride);
                if (messageOverride != null)
                    messageOverride = config.string("messages.join-prefix", "<dark_green>❖ ") + messageOverride;
                String message = config.string("messages.join", "<gray>Игрок <green>%player_name% <gray>присоединился!");
                message = message.replace("%player_name%", player.getUsername());
                message = config.string("messages.join-prefix", "<dark_green>❖ ") + message;
                sendMessageToAll(messageOverride != null ? mm(messageOverride) : mm(message));
            }
        } catch (Exception e) {
            log.error("Error while sending join message", e);
        }
    }

    private void leaveMessage(Player player) {
        if (Velocity.isShuttingDown.get()) return;
        //if (!player.isActive()) return;
        String messageOverride = null;
        if (commonCore.getJoinMessagesRedisRepo() != null) {
            JoinMessages join = commonCore.getJoinMessagesRedisRepo().getOrNull(player.getUsername()).join();
            if (join != null && !join.getLeaveMessages().isEmpty()) {
                int idx = ThreadLocalRandom.current().nextInt(join.getLeaveMessages().size());
                int current = 0;
                for (String leaveMessage : join.getLeaveMessages()) {
                    if (current == idx) {
                        messageOverride = leaveMessage;
                        messageOverride = messageOverride.replace("%player_name%", player.getUsername());
                        break;
                    }
                    current++;
                }
            }
        }
        commonCore.getDiscordBot().sendJoinEmbed(player.getUsername(), DiscordBot.JoinType.LEAVE, messageOverride);
        commonCore.getTelegramBot().sendJoinMessage(player.getUsername(), DiscordBot.JoinType.LEAVE, messageOverride);
        if (messageOverride != null)
            messageOverride = config.string("messages.leave-prefix", "<dark_red>❖ ") + messageOverride;
        String message = config.string("messages.leave", "gray>Игрок <red>%player_name% <gray>вышел!");
        message = message.replace("%player_name%", player.getUsername());
        message = config.string("messages.leave-prefix", "<dark_red>❖ ") + message;
        sendMessageToAll(messageOverride != null ? mm(messageOverride) : mm(message));
    }

}

File: src/main/kotlin/ru/arc/velocity/listeners/JoinListener.kt
package ru.arc.velocity.listeners

import com.velocitypowered.api.event.Subscribe
import com.velocitypowered.api.event.connection.DisconnectEvent
import com.velocitypowered.api.event.connection.PostLoginEvent
import com.velocitypowered.api.event.player.ServerConnectedEvent
import org.slf4j.LoggerFactory
import ru.arc.chat.ChatModeService
import ru.arc.join.AnnouncementPermissions
import ru.arc.join.JoinSessionAnnouncements
import ru.arc.join.VelocityAnnouncementPlayer
import ru.arc.velocity.Velocity

class JoinListener(
    private val announcements: JoinSessionAnnouncements,
) {
    @Subscribe(async = true)
    fun onPlayerAuthenticated(event: PostLoginEvent) {
        if (Velocity.isShuttingDown.get()) return
        val player = event.player
        Velocity.playerActivityTracker?.markSeen(player.uniqueId)
        ChatModeService.track(player.uniqueId)
        Velocity.playerListAnnouncer?.addPlayer(
            player.uniqueId,
            player.username,
            player.currentServer.map { it.serverInfo.name }.orElse(""),
        )
        Velocity.discordBot?.refreshPlayerListFromProxy()
        announcements.onPostLogin(
            VelocityAnnouncementPlayer(player),
            AnnouncementPermissions(
                external = player.hasPermission(EXTERNAL_JOIN_ANNOUNCEMENT_PERMISSION),
            ),
        )
    }

    @Subscribe(async = true)
    fun onPlayerLeave(event: DisconnectEvent) {
        val player = event.player
        announcements.onDisconnect(VelocityAnnouncementPlayer(player))
        if (event.loginStatus == DisconnectEvent.LoginStatus.SUCCESSFUL_LOGIN) {
            Velocity.playerActivityTracker?.markSeen(player.uniqueId)
        }
        ChatModeService.untrack(player.uniqueId)
        Velocity.playerListAnnouncer?.removePlayer(player.uniqueId)
        if (!Velocity.isShuttingDown.get()) {
            Velocity.discordBot?.refreshPlayerListFromProxy()
        }
    }

    @Subscribe(async = true)
    fun onChangeServer(event: ServerConnectedEvent) {
        val server = event.server.serverInfo.name
        val username = event.player.username
        Velocity.playerListAnnouncer?.updatePlayer(event.player.uniqueId, username, server)
        Velocity.discordBot?.refreshPlayerListFromProxy()
        val discord = Velocity.discordBot
        if (discord?.isVerificationBackendAllowed(server) == true) {
            discord.reconcileIdentity(event.player.uniqueId, username).whenComplete { _, error ->
                if (error != null) {
                    log.warn("Discord identity reconciliation failed for {}", username, error)
                }
            }
        }
        Velocity.telegramBot?.takeIf { it.isIdentityBackendAllowed(server) }
            ?.refreshIdentity(event.player.uniqueId, username)
    }

    companion object {
        internal const val EXTERNAL_JOIN_ANNOUNCEMENT_PERMISSION = "arc.join-message.external"
        private val log = LoggerFactory.getLogger(JoinListener::class.java)
    }
}
